File: Sources/CgRPC/shim/channel.c
/*
 * Copyright 2016, gRPC Authors All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "internal.h"
#include "cgrpc.h"
#include <grpc/support/string_util.h>
#include <grpc/support/alloc.h>

#include <stdlib.h>

cgrpc_channel *cgrpc_channel_create(const char *address,
                                    grpc_arg *args,
                                    int num_args) {
  cgrpc_channel *c = (cgrpc_channel *) gpr_zalloc(sizeof (cgrpc_channel));

  grpc_channel_args channel_args;
  channel_args.args = args;
  channel_args.num_args = num_args;

  c->channel = grpc_insecure_channel_create(address, &channel_args, NULL);
  c->completion_queue = grpc_completion_queue_create_for_next(NULL);
  return c;
}

cgrpc_channel *cgrpc_channel_create_secure(const char *address,
                                           const char *pem_root_certs,
                                           const char *client_certs,
                                           const char *client_private_key,
                                           grpc_arg *args,
                                           int num_args) {
  cgrpc_channel *c = (cgrpc_channel *) gpr_zalloc(sizeof (cgrpc_channel));

  grpc_channel_args channel_args;
  channel_args.args = args;
  channel_args.num_args = num_args;

  grpc_ssl_pem_key_cert_pair client_credentials;
  grpc_ssl_pem_key_cert_pair *client_credentials_pointer = NULL;
  if (client_certs != NULL && client_private_key != NULL) {
    client_credentials.cert_chain = client_certs;
    client_credentials.private_key = client_private_key;
    client_credentials_pointer = &client_credentials;
  }
  grpc_channel_credentials *creds = grpc_ssl_credentials_create(pem_root_certs, client_credentials_pointer, NULL);

  c->channel = grpc_secure_channel_create(creds, address, &channel_args, NULL);
  c->completion_queue = grpc_completion_queue_create_for_next(NULL);

  grpc_channel_credentials_release(creds);

  return c;
}

cgrpc_channel *cgrpc_channel_create_google(const char *address,
                                           grpc_arg *args,
                                           int num_args) {
  cgrpc_channel *c = (cgrpc_channel *) gpr_zalloc(sizeof (cgrpc_channel));

  grpc_channel_args channel_args;
  channel_args.args = args;
  channel_args.num_args = num_args;

  grpc_channel_credentials *google_creds = grpc_google_default_credentials_create();

  c->channel = grpc_secure_channel_create(google_creds, address, &channel_args, NULL);
  c->completion_queue = grpc_completion_queue_create_for_next(NULL);

  grpc_channel_credentials_release(google_creds);

  return c;
}


void cgrpc_channel_destroy(cgrpc_channel *c) {
  grpc_channel_destroy(c->channel);
  c->channel = NULL;
  gpr_free(c);
}

cgrpc_call *cgrpc_channel_create_call(cgrpc_channel *channel,
                                      const char *method,
                                      const char *host,
                                      double timeout) {
  // create call
  grpc_slice host_slice = grpc_slice_from_copied_string(host);
  grpc_slice method_slice = grpc_slice_from_copied_string(method);
  gpr_timespec deadline = cgrpc_deadline_in_seconds_from_now(timeout);
  // The resulting call will have a retain call of +1. We'll release it in `cgrpc_call_destroy()`.
  grpc_call *channel_call = grpc_channel_create_call(channel->channel,
                                                     NULL,
                                                     GRPC_PROPAGATE_DEFAULTS,
                                                     channel->completion_queue,
                                                     method_slice,
                                                     &host_slice,
                                                     deadline,
                                                     NULL);
  grpc_slice_unref(host_slice);
  grpc_slice_unref(method_slice);
  cgrpc_call *call = (cgrpc_call *) gpr_zalloc(sizeof(cgrpc_call));
  call->call = channel_call;
  return call;
}

cgrpc_completion_queue *cgrpc_channel_completion_queue(cgrpc_channel *channel) {
  return channel->completion_queue;
}

grpc_connectivity_state cgrpc_channel_check_connectivity_state(cgrpc_channel *channel, int try_to_connect) {
  return grpc_channel_check_connectivity_state(channel->channel, try_to_connect);
}

void cgrpc_channel_watch_connectivity_state(cgrpc_channel *channel, cgrpc_completion_queue *completion_queue, grpc_connectivity_state last_observed_state, double deadline, void *tag) {
  gpr_timespec deadline_seconds = cgrpc_deadline_in_seconds_from_now(deadline);
  return grpc_channel_watch_connectivity_state(channel->channel, last_observed_state, deadline_seconds, completion_queue, tag);
}

File: Sources/SwiftGRPC/Core/Channel.swift
/*
 * Copyright 2016, gRPC Authors All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#if SWIFT_PACKAGE
import CgRPC
#endif
import Foundation

/// Used to hold weak references to objects since `NSHashTable<T>.weakObjects()` isn't available on Linux.
/// If/when this type becomes available on Linux, this should be replaced.
private final class WeakReference<T: AnyObject> {
  private(set) weak var value: T?

  init(value: T) {
    self.value = value
  }
}

/// A gRPC Channel
public class Channel {
  private let mutex = Mutex()
  /// Pointer to underlying C representation
  private let underlyingChannel: UnsafeMutableRawPointer
  /// Completion queue for channel call operations
  private let completionQueue: CompletionQueue
  /// Weak references to API calls using this channel that are in-flight
  private var activeCalls = [WeakReference<Call>]()
  /// Observer for connectivity state changes. Created lazily if needed
  private var connectivityObserver: ConnectivityObserver?
  /// Whether the gRPC channel has been shut down
  private var hasBeenShutdown = false

  /// Timeout for new calls
  public var timeout: TimeInterval = 600.0

  /// Default host to use for new calls
  public var host: String

  /// Errors that may be thrown by the channel
  enum Error: Swift.Error {
    /// Action cannot be performed because the channel has already been shut down
    case alreadyShutdown
    /// Failed to create a new call within the gRPC stack
    case callCreationFailed
  }

  /// Initializes a gRPC channel
  ///
  /// - Parameter address: the address of the server to be called
  /// - Parameter secure: if true, use TLS
  /// - Parameter arguments: list of channel configuration options
  public convenience init(address: String, secure: Bool = true, arguments: [Argument] = []) {
    gRPC.initialize()

    let argumentWrappers = arguments.map { $0.toCArg() }
    self.init(host: address, underlyingChannel: withExtendedLifetime(argumentWrappers) {
      var argumentValues = argumentWrappers.map { $0.wrapped }
      if secure {
        return cgrpc_channel_create_secure(address, kRootCertificates, nil, nil, &argumentValues, Int32(arguments.count))
      } else {
        return cgrpc_channel_create(address, &argumentValues, Int32(arguments.count))
      }
    })
  }

  /// Initializes a gRPC channel
  ///
  /// - Parameter address: the address of the server to be called
  /// - Parameter arguments: list of channel configuration options
  public convenience init(googleAddress: String, arguments: [Argument] = []) {
    gRPC.initialize()

    let argumentWrappers = arguments.map { $0.toCArg() }
    self.init(host: googleAddress, underlyingChannel: withExtendedLifetime(argumentWrappers) {
      var argumentValues = argumentWrappers.map { $0.wrapped }
      return cgrpc_channel_create_google(googleAddress, &argumentValues, Int32(arguments.count))
    })
  }

  /// Initializes a gRPC channel
  ///
  /// - Parameter address: the address of the server to be called
  /// - Parameter certificates: a PEM representation of certificates to use.
  /// - Parameter clientCertificates: a PEM representation of the client certificates to use
  /// - Parameter clientKey: a PEM representation of the client key to use
  /// - Parameter arguments: list of channel configuration options
  public convenience init(address: String, certificates: String = kRootCertificates, clientCertificates: String? = nil, clientKey: String? = nil, arguments: [Argument] = []) {
    gRPC.initialize()

    let argumentWrappers = arguments.map { $0.toCArg() }
    self.init(host: address, underlyingChannel: withExtendedLifetime(argumentWrappers) {
      var argumentValues = argumentWrappers.map { $0.wrapped }
      return cgrpc_channel_create_secure(address, certificates, clientCertificates, clientKey, &argumentValues, Int32(arguments.count))
    })
  }

  /// Shut down the channel. No new calls may be made using this channel after it is shut down. Any in-flight calls using this channel will be canceled
  public func shutdown() {
    self.mutex.synchronize {
      guard !self.hasBeenShutdown else { return }

      self.hasBeenShutdown = true
      self.connectivityObserver?.shutdown()
      cgrpc_channel_destroy(self.underlyingChannel)
      self.completionQueue.shutdown()
      self.activeCalls.forEach { $0.value?.cancel() }
    }
  }

  deinit {
    self.shutdown()
  }

  /// Constructs a Call object to make a gRPC API call
  ///
  /// - Parameter method: the gRPC method name for the call
  /// - Parameter host: the gRPC host name for the call. If unspecified, defaults to the Client host
  /// - Parameter timeout: a timeout value in seconds
  /// - Returns: a Call object that can be used to perform the request
  public func makeCall(_ method: String, host: String? = nil, timeout: TimeInterval? = nil) throws -> Call {
    self.mutex.lock()
    defer { self.mutex.unlock() }

    guard !self.hasBeenShutdown else {
      throw Error.alreadyShutdown
    }

    guard let underlyingCall = cgrpc_channel_create_call(
      self.underlyingChannel, method, host ?? self.host, timeout ?? self.timeout)
      else { throw Error.callCreationFailed }

    let call = Call(underlyingCall: underlyingCall, owned: true, completionQueue: self.completionQueue)
    self.activeCalls.append(WeakReference(value: call))
    return call
  }

  /// Check the current connectivity state
  ///
  /// - Parameter tryToConnect: boolean value to indicate if should try to connect if channel's connectivity state is idle
  /// - Returns: a ConnectivityState value representing the current connectivity state of the channel
  public func connectivityState(tryToConnect: Bool = false) -> ConnectivityState {
    return ConnectivityState(cgrpc_channel_check_connectivity_state(underlyingChannel, tryToConnect ? 1 : 0))
  }

  /// Subscribe to connectivity state changes
  ///
  /// - Parameter callback: block executed every time a new connectivity state is detected
  public func addConnectivityObserver(callback: @escaping (ConnectivityState) -> Void) {
    self.mutex.synchronize {
      let observer: ConnectivityObserver
      if let existingObserver = self.connectivityObserver {
        observer = existingObserver
      } else {
        observer = ConnectivityObserver(underlyingChannel: self.underlyingChannel)
        self.connectivityObserver = observer
      }

      observer.addConnectivityObserver(callback: callback)
    }
  }

  // MARK: - Private

  private init(host: String, underlyingChannel: UnsafeMutableRawPointer) {
    self.host = host
    self.underlyingChannel = underlyingChannel
    self.completionQueue = CompletionQueue(underlyingCompletionQueue: cgrpc_channel_completion_queue(underlyingChannel),
                                           name: "Client")

    self.completionQueue.run()
    self.scheduleActiveCallCleanUp()
  }

  private func scheduleActiveCallCleanUp() {
    DispatchQueue.global(qos: .background).asyncAfter(deadline: .now() + 10.0) { [weak self] in
      self?.cleanUpActiveCalls()
    }
  }

  private func cleanUpActiveCalls() {
    self.mutex.synchronize {
      self.activeCalls = self.activeCalls.filter { $0.value != nil }
    }
    self.scheduleActiveCallCleanUp()
  }
}
